lib/pengine/rules.c

```diff
@@ -171,8 +171,23 @@ populate_hash(xmlNode * nvpair_list, GHashTable * hash, gboolean overwrite, xmlN
                     g_hash_table_remove(hash, name);
                 }
                 continue;
+            }
+
+            if (pcmk__str_eq(value, PCMK__ROLE_PROMOTED_LEGACY, pcmk__str_casei)) {
+                pcmk__warn_once(pcmk__wo_master_role,
+                                "Support for the " PCMK__ROLE_PROMOTED_LEGACY
+                                " role is deprecated and will be removed in a"
+                                "future release. Use " PCMK__ROLE_PROMOTED
+                                " instead.");
+            } else if (pcmk__str_eq(value, PCMK__ROLE_UNPROMOTED_LEGACY, pcmk__str_casei)) {
+                pcmk__warn_once(pcmk__wo_slave_role,
+                                "Support for the " PCMK__ROLE_UNPROMOTED_LEGACY
+                                " role is deprecated and will be removed in a"
+                                "future release. Use " PCMK__ROLE_UNPROMOTED
+                                " instead.");
+            }
```

Comment:
I don't *think* this should be necessary; the rule code uses `pcmk_parse_role()` too. However it's only called when needed, and the schema currently allows role in places it will never be used, so that might be why it doesn't always show up. Still, I'd rather not do it here unless we check the name too.